tests/integration/queries/test_migration_progress.py

```diff
@@ -0,0 +1,353 @@
+    for cluster in (
+        ClusterInfo("1", success=1, failures=""),
+        ClusterInfo("2", success=0, failures="Uses azure service principal credentials config in cluster"),
+    ):
+        record = (
+            "clusters",
+            [cluster.cluster_id],
+            cluster,
+            cluster.failures.split("\n") if cluster.failures else [],
+            "Cor",
+        )
+        records.append(record)
+    return records
+
+
+@pytest.fixture
+def pipelines():
+    records = []
+    for pipeline in (
+        PipelineInfo("1", success=1, failures=""),
+        PipelineInfo(
+            "2", success=0, failures="Uses passthrough config: spark.databricks.passthrough.enabled in pipeline"
+        ),
+    ):
+        record = (
+            "pipelines",
+            [pipeline.pipeline_id],
+            pipeline,
+            pipeline.failures.split("\n") if pipeline.failures else [],
+            "Cor",
+        )
+        records.append(record)
+    return records
+
+
+@pytest.fixture
+def policies():
+    records = []
+    for policy in (
+        PolicyInfo("1", "policy1", success=1, failures=""),
+        PolicyInfo(
+            "2",
+            "policy2",
+            success=0,
+            failures="Uses azure service principal credentials config in policy",
+        ),
+    ):
+        record = (
+            "policies",
+            [policy.policy_id],
+            policy,
+            policy.failures.split("\n") if policy.failures else [],
+            "Cor",
+        )
+        records.append(record)
+    return records
+
+
+@pytest.fixture
+def historical_objects(
+    tables,
+    table_migration_statuses,
+    udfs,
+    grants,
+    jobs,
+    clusters,
+    pipelines,
+    policies,
+):
+    return tables + table_migration_statuses + udfs + grants + jobs + clusters + pipelines + policies
+
+
+@pytest.fixture
+def catalog_populated(
+    installation_ctx: MockInstallationContext,
+    workflow_runs: list[WorkflowRun],
+    historical_objects,
+) -> str:
+    """Populate the UCX catalog given the objects from the fixtures.
+
+    For optimization purposes, this fixture could be "module" (or "session") scoped. However, dependant fixtures are
+    "function" scoped, thus one should first evaluate if those can be changed.
+    """
+    installation_ctx.progress_tracking_installation.run()
+    ucx_catalog = installation_ctx.ucx_catalog
+    installation_ctx.sql_backend.save_table(
+        f"{ucx_catalog}.multiworkspace.workflow_runs", workflow_runs, WorkflowRun, mode="overwrite"
+    )
+    workspace_id = installation_ctx.workspace_client.get_workspace_id()
+    historicals = []
+    for job_run_id in range(1, 3):  # No changes between migration progress run
+        for table_name, id_, instance, failures, owner in historical_objects:
+            # TODO: Use historical encoder from https://github.com/databrickslabs/ucx/pull/2743/
+            data = {
+                field.name: str(getattr(instance, field.name))
```

Comment:
https://github.com/databrickslabs/pylint-plugin?tab=readme-ov-file#r8923-rewrite-as-for-loop

Reply:
Kept it to replace it with the encoder from Andrew's PR